akd_local_auditor/src/storage/s3.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree and the Apache
+// License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+// of this source tree.
+
+//! This module comprises S3 bucket READ ONLY access to download and parse
+//! Audit Proofs
+
+use anyhow::{bail, Result};
+use async_trait::async_trait;
+use aws_config::RetryConfig;
+use aws_sdk_s3 as s3;
+use clap::Args;
+use log::{debug, error};
+use std::convert::TryInto;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+const MIN_BUCKET_CHARS: usize = 3;
+const MAX_BUCKET_CHARS: usize = 63;
+const ALLOWED_BUCKET_CHARS: [char; 38] = [
+    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's',
+    't', 'u', 'v', 'w', 'x', 'y', 'z', '.', '-', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+];
+
+fn is_bucket_name_valid(s: &str) -> Result<String, String> {
```

Comment:
I assume that we're aiming to implement a portion of the [rules for S3 bucket names](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html) as validation here; is that correct? If so, do you think that's worth trying to keep up? I ask, because a user of the tool should/will eventually see an error from AWS in any case when we attempt to read from the bucket. Additionally, if bucket naming conventions change, then there's potential that the validation here becomes stale. I'm less concerned about the latter, as I don't see bucket names being named oddly often.

Reply:
It's more that we won't even try and make a request unless it matches the _basics_ of naming conventions which is 3-63 chars, lower-case a-z, 0-9, and/or ., -

That's all it is and it hasn't changed in a long time. I'm not saying we should go full-spec match across the board, but this is an easy check to see if they mistyped the name of the bucket. They're URL names afterall so there's already a tight limit on what's possible